protocol/install.sh

```diff
@@ -1,17 +1,165 @@
+
+  --clean                  Delete any previously installed header
+                           files from ${INSTALL_DESTINATION}/include and any
+                           shared library files from ${INSTALL_DESTINATION}/lib.
+
+  --cleandir [path]        Delete any previously installed header
+                           and library files from the specified path.
+
+  -h, --help               Show this help message."
+
+function make_bindings() {
+  cd "${PROTOBUF_DIR}" && make -j all || exit 1
+}
+
+function clean_bindings() {
+  cd "${PROTOBUF_DIR}" && make clean
+}
+
+function install_protobuf() {
+  echo ">>> INSTALLING PROTOBUF DEPENDENCIES"
+  apt-get update && apt-get install "${AUTO_INSTALL}" autoconf automake libtool curl make g++ unzip || exit 1
+
+  mkdir -p "${SCRIPT_DIR}"/install
+  cd "${SCRIPT_DIR}"/install || exit 1
+  wget -O protobuf-cpp-"${PROTOBUF_VERSION}".tar.gz \
+    https://github.com/protocolbuffers/protobuf/releases/download/v"${PROTOBUF_VERSION}"/protobuf-cpp-"${PROTOBUF_VERSION}".tar.gz &&
+    tar -xzf protobuf-cpp-"${PROTOBUF_VERSION}".tar.gz &&
+    rm protobuf-cpp-"${PROTOBUF_VERSION}".tar.gz
+
+  cd "${SCRIPT_DIR}"/install/protobuf-"${PROTOBUF_VERSION}" || exit 1
+  ./autogen.sh && ./configure && make && make install || exit 1
+  ldconfig
+}
+
+function install_state_representation() {
+  echo ">> INSTALLING CONTROL LIBRARY DEPENDENCIES"
+  mkdir -p "${SCRIPT_DIR}"/install
+  cd "${SCRIPT_DIR}"/install || exit 1
+  CL_INSTALL_SCRIPT="$(dirname "${SCRIPT_DIR}")"/source/install.sh
+  if [ -f "$CL_INSTALL_SCRIPT" ]; then
+    bash "${CL_INSTALL_SCRIPT}" --no-controllers --no-dynamical-systems --no-robot-model "${AUTO_INSTALL}"
+  else
+    echo ">>> INSTALL SCRIPT NOT FOUND: ${CL_INSTALL_SCRIPT}!"
+    exit 1
+  fi
+  ldconfig
+}
+
+function install_clproto() {
+  cd "${CLPROTO_DIR}" && mkdir -p build && cd build || exit 1
+
+  cmake -DCMAKE_BUILD_TYPE=Release \
+    -DCMAKE_INSTALL_PREFIX="${INSTALL_DESTINATION}" .. || exit 1
+
+  make -j && make install || exit 1
+  ldconfig
+}
+
+function uninstall() {
+  function delete_components() {
+    rm -r "${INSTALL_DESTINATION}"/include/clproto
+    rm -r "${INSTALL_DESTINATION}"/lib/libclproto*.so
+    clean_bindings
+  }
+
+  delete_components >/dev/null 2>&1
+
+  echo "Deleted any clproto artefacts from ${INSTALL_DESTINATION} and ${PROTOBUF_DIR}."
+}
+
+while [ "$#" -gt 0 ]; do
+  case "$1" in
+  -y | --auto)
+    AUTO_INSTALL="-y"
+    shift 1
+    ;;
+  -d | --dir)
+    INSTALL_DESTINATION=$2
+    shift 2
+    ;;
+  --bindings-only)
+    BINDINGS_ONLY=true
+    shift 1
+    ;;
+  --clean-bindings)
+    clean_bindings
+    exit 0
+    ;;
+  --clean)
+    uninstall
+    exit 0
+    ;;
+  --cleandir)
+    INSTALL_DESTINATION=$2
+    uninstall
+    exit 0
+    ;;
+  -h | --help)
+    echo "$HELP_MESSAGE"
+    exit 0
+    ;;
+
+  -*)
+    echo "Unknown option: $1" >&2
+    echo "$FAIL_MESSAGE"
+    exit 1
+    ;;
+  esac
+done
+
+if [ -x "$(command -v protoc)" ]; then
```

Comment:
I couldn't find the docs for `-x` but this is wrong as it is now. Either put the `if ! ` or change the `-x` option.

```suggestion
if ! [ -x "$(command -v protoc)" ]; then
```

Reply:
Huh ok, interesting if this failed for you. I tested the command -v part on a container when I made this and it worked fine. The -x test condition also is standard in all shell interpreters to see if a file exists and is executable. Maybe you were googling it with the literal "bash -x"? As you probably know, that tells google to omit results that contain x

https://www.gnu.org/software/bash/manual/html_node/Bash-Conditional-Expressions.html

Reply:
I tested it inside a container too and it didn't work. See https://github.com/aica-technology/docker-images/blob/feature/cl-ros-images/cl_ros_ws/Dockerfile for an example. I'll try again

Reply:
if you checkout `feature/cl-ros-images` and try to do `build.sh --cl-branch feature/revise-install` in the cl_ros_ws directory, it installs protoc again, but if you do `build.sh --cl-branch fix/revise-install` it works.

Reply:
oh yes yes of course, I see now what you meant originally. I thought your issue was with the -x command not working, but obviously it's just that the logical statement is inverted. If it's _not_ an executable file, do the following. My mistake.